src/Gui/Tree.cpp

```diff
@@ -2399,21 +2416,33 @@ bool TreeWidget::dropInObject(QDropEvent* event, TargetItemInfo& targetInfo,
         info.subs.swap(v.second);
 
         // check if items can be dragged
-        if (da == Qt::MoveAction &&
-            item->myOwner == targetItemObj->myOwner &&
-            vp->canDragAndDropObject(item->object()->getObject()))
-        {
-            // check if items can be dragged
+        if (da == Qt::MoveAction && item->myOwner == targetItemObj->myOwner && vp->canDragAndDropObject(obj)) {
             auto parentItem = item->getParentItem();
-            if (!parentItem)
-                info.dragging = true;
-            else if (parentItem->object()->canDragObjects()
-                && parentItem->object()->canDragObject(item->object()->getObject()))
-            {
+            if (!parentItem) {
                 info.dragging = true;
+            }
+            else {
                 auto vpp = parentItem->object();
-                info.parent = vpp->getObject()->getNameInDocument();
-                info.parentDoc = vpp->getObject()->getDocument()->getName();
+                // We querry all the parents recursively.
+                bool allParentsOK = true;
+                while (parentItem) {
+                    Base::Console().Warning("parentItem %s\n", parentItem->getName());
+                    if (!parentItem->object()->canDragObjectToTarget(obj, targetObj)) {
+                        allParentsOK = false;
+                        break;
+                    }
+                    parentItem = parentItem->getParentItem();
+                }

```

Comment:
This seems like repeated code (even with typo in the comment!) and hence should be refactored into its own method.

Reply:
Done